### PR TITLE
[DebugInfo][LICM] Fix missing debug location updates

### LINKED ISSUE #91716: [DebugInfo][LICM] Missing debug location updates
LICM[-L932](https://github.com/llvm/llvm-project/blob/6aac30fa43f094ac25269bda163dc89a88cb8da7/llvm/lib/Transforms/Scalar/LICM.cpp#L932), [-L937](https://github.com/llvm/llvm-project/blob/6aac30fa43f094ac25269bda163dc89a88cb8da7/llvm/lib/Transforms/Scalar/LICM.cpp#L937): Newly created FDiv `ReciprocalDivisor` is inserted before the instruction `I` replaced by new FMul `Product`.
```cpp
->  auto ReciprocalDivisor = BinaryOperator::CreateFDiv(One, Divisor);
|   ReciprocalDivisor->setFastMathFlags(I.getFastMathFlags());
|   SafetyInfo->insertInstructionTo(ReciprocalDivisor, I.getParent());
|_  ReciprocalDivisor->insertBefore(&I);

->  auto Product =
|     BinaryOperator::CreateFMul(I.getOperand(0), ReciprocalDivisor);
|   Product->setFastMathFlags(I.getFastMathFlags());
|   SafetyInfo->insertInstructionTo(Product, I.getParent());
|   Product->insertAfter(&I);
|   I.replaceAllUsesWith(Product);
|_  eraseInstruction(I, *SafetyInfo, MSSAU);
```

I will give the patch later.

### PR DESCRIPTION
**Pull request by Apochens**
Fix #91716.
* [LICM-L937](https://github.com/llvm/llvm-project/blob/6aac30fa43f094ac25269bda163dc89a88cb8da7/llvm/lib/Transforms/Scalar/LICM.cpp#L937) is tested by `preserving-debugloc-fmul.ll`.

For [LICM-L932](https://github.com/llvm/llvm-project/blob/6aac30fa43f094ac25269bda163dc89a88cb8da7/llvm/lib/Transforms/Scalar/LICM.cpp#L932) (`ReciprocalDivisor`), I added the debug location update. However, `ReciprocalDivisor` would be regarded as a loop invariant in the same function after the creation:
```cpp
auto ReciprocalDivisor = BinaryOperator::CreateFDiv(One, Divisor);      // Created here
ReciprocalDivisor->setFastMathFlags(I.getFastMathFlags());
SafetyInfo->insertInstructionTo(ReciprocalDivisor, I.getParent());
ReciprocalDivisor->insertBefore(&I);

auto Product =
    BinaryOperator::CreateFMul(I.getOperand(0), ReciprocalDivisor);
Product->setFastMathFlags(I.getFastMathFlags());
SafetyInfo->insertInstructionTo(Product, I.getParent());
Product->insertAfter(&I);
I.replaceAllUsesWith(Product);
eraseInstruction(I, *SafetyInfo, MSSAU);

hoist(*ReciprocalDivisor, DT, CurLoop, CFH.getOrCreateHoistedBlock(BB),  // Hoisted here
    SafetyInfo, MSSAU, SE, ORE);
```
So, it's debug location would be dropped immediately in the function `hoist()`, and I didn't create a test for it. However, IMO, assigning a proper debug location for a new instruction is reasonable and appropriate, though the assigned debug location would be manipulated afterwards. I'd like to take any suggestion on this special case.

### CODE REVIEW

File: llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll [hunk 1]
```diff
@@ -0,0 +1,78 @@
+; RUN: opt -passes=licm -verify-memoryssa -S < %s | FileCheck %s
+
+; Test that LICM's hoistRegin()
```
Comment (OCHyams) [inline llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll:3]: Is this comment unfinished?

File: llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+!12 = !DILocalVariable(name: "3", scope: !5, file: !1, line: 4, type: !13)
+!13 = !DIBasicType(name: "ty64", size: 64, encoding: DW_ATE_unsigned)
+!14 = !DILocalVariable(name: "4", scope: !5, file: !1, line: 5, type: !13)
+!15 = !DILocalVariable(name: "5", scope: !5, file: !1, line: 6, type: !13)
+!16 = !DILocalVariable(name: "6", scope: !5, file: !1, line: 7, type: !13)
+!17 = !DILocalVariable(name: "7", scope: !5, file: !1, line: 8, type: !13)
+!18 = !DILocalVariable(name: "8", scope: !5, file: !1, line: 9, type: !10)
+!19 = !DILocalVariable(name: "9", scope: !5, file: !1, line: 10, type: !10)
+!20 = !DILocalVariable(name: "10", scope: !5, file: !1, line: 11, type: !10)
+!21 = !DILocalVariable(name: "11", scope: !5, file: !1, line: 12, type: !22)
```
Comment (OCHyams) [inline llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll:59]: if you give opt the option `--debugify-level=locations` then the debugify pass only adds `!dbg` metadata and doesn't insert `dbg.value`s or create DILocalVariables.

In any case, please can you remove these unused variables?

You could also remove remove `!dbg` attachments and the `!DILocations` other than the one for `line: 5`, as that's the only one used in the test, but this isn't super important (feel free to ignore).

**Review by OCHyams — COMMENTED** [review]
Thanks for another patch. Code and test SGTM, but there are a couple of questions first.

> So, it's debug location would be dropped immediately in the function hoist(), and I didn't create a test for it. However, IMO, assigning a proper debug location for a new instruction is reasonable and appropriate, though the assigned debug location would be manipulated afterwards. I'd like to take any suggestion on this special case.

This is an interesting observation, that `hoist` calls `updateLocationAfterHoist` which calls `dropLocation`.

I think there's a future in which we preserve more location information even after hoists - this is an area we're actively looking into right now (cc @SLTozer).

Given that, I would actually be in favour of doing the location propagation as you have done, and also adding an assert after the `hoist` which check that the location has been dropped (I think `assert(!ReciprocalDivisor.getDebugLoc())` should work? ), with a comment pointing out that if the assert fires a test case should be added and it's safe to remove the assert. 

This may end up being an unpopular opinion (propagating the location is a noop because it'll get dropped, but it still has a runtime cost). @pogo59 wdyt? I might be making it up but I think I recall a similar discussion happening recently where you were involved.

**Comment by pogo59** [comment]
> This may end up being an unpopular opinion (propagating the location is a noop because it'll get dropped, but it still has a runtime cost). @pogo59 wdyt? I might be making it up but I think I recall a similar discussion happening recently where you were involved.

I vaguely remember that, but not in enough detail to be sure I will say something consistent here.

It would be an unusual use of `assert` for sure, and I think it would be preferable to have a test that checks the location is lost. Then the test is already written, it will fail when we handle line locations differently, and a comment in the test can say we expect this to need updating when we fix hoist to do the right thing.

**Comment by OCHyams** [comment]
> It would be an unusual use of assert for sure, and I think it would be preferable to have a test that checks the location is lost. Then the test is already written, it will fail when we handle line locations differently, and a comment in the test can say we expect this to need updating when we fix hoist to do the right thing.

+1, that makes a lot more sense.

**Comment by Apochens** [comment]
> It would be an unusual use of `assert` for sure, and I think it would be preferable to have a test that checks the location is lost. Then the test is already written, it will fail when we handle line locations differently, and a comment in the test can say we expect this to need updating when we fix hoist to do the right thing.

Actually, the `fdiv` with debug location dropped can be tested by the same test case (ie., `preserving-debugloc-fmul.ll`). I can add checks for that in this test and a comment pointing out this special case. Would that be acceptable?

File: llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll (hunk 1 above)
Comment (Apochens) [reply llvm/test/Transforms/LICM/preversing-debugloc-fmul.ll:3]: Oh, that must be a commit mistake. I'll fix it.

**Comment by OCHyams** [comment]
> Would that be acceptable?

Sounds good, yes please

**Review by OCHyams — APPROVED** [review]
The comment in the test is very helpful - all LGTM, thanks.

**Review by jryans — APPROVED** [review]
Looks good to me as well! 😄 

I'll go ahead and merge this now.

**Pull request merged**